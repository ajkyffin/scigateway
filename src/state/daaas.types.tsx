import { Step } from 'react-joyride';

export const AuthSuccessType = 'daaas:auth_success';
export const AuthFailureType = 'daaas:auth_failure';
export const LoadingAuthType = 'daaas:loading_auth';
export const LoadAuthProviderType = 'daaas:loading_auth_provider';
export const TokenExpiredType = 'daaas:expired_token';
export const ConfigureStringsType = 'daaas:configure_strings';
export const LoginType = 'daaas:login';
export const NotificationType = 'daaas:api:notification';
export const RegisterRouteType = 'daaas:api:register_route';
export const RequestPluginRerenderType = 'daaas:api:plugin_rerender';
export const SignOutType = 'daaas:signout';
export const ToggleDrawerType = 'daaas:toggledrawer';
export const DismissNotificationType = 'daaas:dismissnotification';
export const ConfigureFeatureSwitchesType = 'daaas:feature_switches';
export const AddNotificationsType = 'daaas:add_notification';
export const SiteLoadingType = 'daaas:site_loading';
export const ConfigureAnalyticsType = 'daaas:configure_analytics';
export const InitialiseAnalyticsType = 'daaas:initialise_analytics';
export const ToggleHelpType = 'daaas:toggle_help';
export const AddHelpTourStepsType = 'daaas:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
