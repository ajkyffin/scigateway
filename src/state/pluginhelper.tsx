import {
  GroupedPlugins,
  PluginConfig,
  RegisterRoutePayload,
} from './daaas.types';

// Custom comparer for plugin objects, sorts keys by order (asc), then by name (asc)
export const comparePlugins = (a: PluginConfig, b: PluginConfig): number =>
  a.order === b.order
    ? a.displayName.localeCompare(b.displayName)
    : a.order - b.order;

function buildMenuItemDictionary(plugins: PluginConfig[]): GroupedPlugins {
  const dict: GroupedPlugins = {};

  plugins.forEach(p => {
    if (!(p.section in dict)) {
      dict[p.section] = [];
    }
    dict[p.section].push(p);
  });

  return dict;
}

// Convert the list of plugins into a structured dataset for rendering
export function structureMenuData(plugins: PluginConfig[]): GroupedPlugins {
  const menuItems = buildMenuItemDictionary(plugins);

  Object.keys(menuItems).forEach(
    section => (menuItems[section] = menuItems[section].sort(comparePlugins))
  );
  return menuItems;
}

// Convert a plugin registration request body into a pluginConfig object
export function buildPluginConfig(payload: RegisterRoutePayload): PluginConfig {
  return { ...payload };
}
