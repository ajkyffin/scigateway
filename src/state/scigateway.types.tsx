import { Step } from 'react-joyride';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadAuthUrlType = 'scigateway:loading_auth_url';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
