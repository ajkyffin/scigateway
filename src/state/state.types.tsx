import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Step } from 'react-joyride';
import {
  ApplicationStrings,
  PluginConfig,
  FeatureSwitches,
} from './scigateway.types';
import { RouterState } from 'connected-react-router';

export interface Plugin {
  name: string;
  src: string;
  enable: boolean;
  location: 'main' | 'left' | 'right';
}

export interface ScigatewayNotification {
  message: string;
  severity: string;
}

export interface ScigatewayState {
  notifications: ScigatewayNotification[];
  plugins: PluginConfig[];
  drawerOpen: boolean;
  siteLoading: boolean;
  showHelp: boolean;
  helpSteps: Step[];
  authorisation: AuthState;
  res?: ApplicationStrings;
  features: FeatureSwitches;
  analytics?: AnalyticsState;
}

export interface StateType {
  scigateway: ScigatewayState;
  router: RouterState;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export interface User {
  username: string;
  avatarUrl: string;
}

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ICATAuthenticator {
  mnemonic: string;
  keys: { name: string; hide?: boolean }[];
  friendly?: string;
  admin?: boolean;
}

export interface AuthProvider {
  isLoggedIn: () => boolean;
  logOut: () => void;
  logIn: (username: string, password: string) => Promise<void>;
  verifyLogIn: () => Promise<void>;
  refresh: () => Promise<void>;
  redirectUrl: string | null;
  user: User | null;
  mnemonic?: string;
  autoLogin?: () => Promise<void>;
}

export interface AuthState {
  failedToLogin: boolean;
  signedOutDueToTokenInvalidation: boolean;
  loading: boolean;
  provider: AuthProvider;
}

export interface AnalyticsState {
  id: string;
  initialised: boolean;
}
