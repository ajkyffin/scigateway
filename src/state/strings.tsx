import { AppStrings } from './scigateway.types';
import { StateType } from './state.types';

export function getAppStrings(
  state: StateType,
  section: string
): AppStrings | undefined {
  return state.scigateway.res ? state.scigateway.res[section] : undefined;
}

export const getString = (res: AppStrings | undefined, key: string): string =>
  (res && res[key]) || key;
