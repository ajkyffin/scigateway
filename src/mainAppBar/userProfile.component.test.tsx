import React from 'react';
import UserProfileComponent from './userProfile.component';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import { StateType } from '../state/state.types';
import configureStore from 'redux-mock-store';
import { initialState } from '../state/reducers/daaas.reducer';
import { Provider } from 'react-redux';
import { push } from 'connected-react-router';
import { MenuItem, Avatar } from '@material-ui/core';
import thunk from 'redux-thunk';
import TestAuthProvider from '../authentication/testAuthProvider';

describe('User profile component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'div' });
    mount = createMount();

    mockStore = configureStore([thunk]);
    state = JSON.parse(JSON.stringify({ daaas: initialState }));
    state.daaas.authorisation.provider = new TestAuthProvider('test-token');
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders sign in button if not signed in', () => {
    state.daaas.authorisation.provider = new TestAuthProvider(null);

    const wrapper = shallow(<UserProfileComponent store={mockStore(state)} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('redirects to login when sign in is pressed', () => {
    state.daaas.authorisation.provider = new TestAuthProvider(null);

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <UserProfileComponent />
      </Provider>
    );

    wrapper
      .find('button')
      .first()
      .simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(push('/login'));
  });

  it('renders default avatar if signed in', () => {
    const wrapper = shallow(<UserProfileComponent store={mockStore(state)} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders user avatar if signed in with avatar url', () => {
    state.daaas.authorisation.provider.user = {
      username: 'test',
      avatarUrl: 'test_url',
    };
    const wrapper = shallow(<UserProfileComponent store={mockStore(state)} />);
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('opens menu when button clicked', () => {
    state.daaas.authorisation.provider.user = {
      username: 'test',
      avatarUrl: 'test_url',
    };
    const wrapper = mount(<UserProfileComponent store={mockStore(state)} />);

    expect(
      wrapper
        .find('#simple-menu')
        .first()
        .prop('open')
    ).toBeFalsy();

    wrapper.find(Avatar).simulate('click');

    expect(
      wrapper
        .find('#simple-menu')
        .first()
        .prop('open')
    ).toBeTruthy();
  });

  it('signs out if sign out clicked', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <UserProfileComponent />
      </Provider>
    );

    wrapper.find('button').simulate('click');
    wrapper.find(MenuItem).simulate('click');

    expect(testStore.getActions().length).toEqual(2);
    expect(testStore.getActions()[0]).toEqual({ type: 'daaas:signout' });
    expect(testStore.getActions()[1]).toEqual(push('/'));
  });
});
