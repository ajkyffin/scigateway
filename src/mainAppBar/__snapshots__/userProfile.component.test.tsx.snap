// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User profile component renders default avatar if signed in 1`] = `
<div
  className="tour-user-profile"
>
  <div>
    <WithStyles(ForwardRef(IconButton))
      aria-label="Open user menu"
      className="UserProfileComponent-button-1"
      onClick={[Function]}
    >
      <AccountCircleIcon />
    </WithStyles(ForwardRef(IconButton))>
    <WithStyles(ForwardRef(Menu))
      anchorEl={null}
      id="simple-menu"
      onClose={[Function]}
      open={false}
    >
      <div
        className="UserProfileComponent-usernameContainer-2"
      >
        <WithStyles(ForwardRef(Typography))>
          Signed in as:
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          className="UserProfileComponent-username-3"
        >
          anonymous
        </WithStyles(ForwardRef(Typography))>
      </div>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(MenuItem))
        aria-label="Manage cookies item"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemIcon))>
          <SettingsIcon />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="manage-cookies-button"
        />
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(MenuItem))
        aria-label="Sign out item"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemIcon))>
          <ExitToAppIcon />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="logout-button"
        />
      </WithStyles(ForwardRef(MenuItem))>
    </WithStyles(ForwardRef(Menu))>
  </div>
</div>
`;

exports[`User profile component renders sign in button if not signed in 1`] = `
<div
  className="tour-user-profile"
>
  <WithStyles(ForwardRef(Button))
    className="UserProfileComponent-button-1"
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(Typography))
      color="inherit"
      noWrap={true}
      style={
        Object {
          "marginTop": 3,
        }
      }
    >
      login-button
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Button))>
</div>
`;

exports[`User profile component renders user avatar if signed in with avatar url 1`] = `
<div
  className="tour-user-profile"
>
  <div>
    <WithStyles(ForwardRef(Avatar))
      alt="user"
      aria-label="Open user menu"
      className="UserProfileComponent-avatar-4"
      onClick={[Function]}
      src="test_url"
    />
    <WithStyles(ForwardRef(Menu))
      anchorEl={null}
      id="simple-menu"
      onClose={[Function]}
      open={false}
    >
      <div
        className="UserProfileComponent-usernameContainer-2"
      >
        <WithStyles(ForwardRef(Typography))>
          Signed in as:
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          className="UserProfileComponent-username-3"
        >
          test
        </WithStyles(ForwardRef(Typography))>
      </div>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(MenuItem))
        aria-label="Manage cookies item"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemIcon))>
          <SettingsIcon />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="manage-cookies-button"
        />
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(MenuItem))
        aria-label="Sign out item"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemIcon))>
          <ExitToAppIcon />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="logout-button"
        />
      </WithStyles(ForwardRef(MenuItem))>
    </WithStyles(ForwardRef(Menu))>
  </div>
</div>
`;
